outlines/models/gemini.py

```diff
@@ -0,0 +1,121 @@
+"""Integration with Gemini's API."""
+from enum import EnumMeta
+from functools import singledispatchmethod
+from types import NoneType
+from typing import Optional, Union
+
+from pydantic import BaseModel
+from typing_extensions import _TypedDictMeta  # type: ignore
+
+from outlines.prompts import Vision
+from outlines.types import Choice, Json
+
+__all__ = ["Gemini"]
+
+
+class GeminiBase:
+    """Base class for the Gemini clients.
+
+    `GeminiBase` is responsible for preparing the arguments to Gemini's
+    `generate_contents` methods: the input (prompt and possibly image), as well
+    as the output type (only JSON).
+
+    """
+
+    @singledispatchmethod
+    def format_input(self, model_input):
+        """Generate the `messages` argument to pass to the client.
+
+        Argument
+        --------
+        model_input
+            The input passed by the user.
+
+        """
+        raise NotImplementedError(
+            f"The input type {input} is not available with Gemini. The only available types are `str` and `Vision`."
+        )
+
+    @format_input.register(str)
+    def format_str_input(self, model_input: str):
+        """Generate the `messages` argument to pass to the client when the user
+        only passes a prompt.
+
+        """
+        return {"contents": [model_input]}
+
+    @format_input.register(Vision)
+    def format_vision_input(self, model_input: Vision):
+        """Generate the `messages` argument to pass to the client when the user
+        passes a prompt and an image.
+
+        """
+        return {"contents": [model_input.prompt, model_input.image]}
+
+    @singledispatchmethod
+    def format_output_type(self, output_type):
+        if output_type.__origin__ == list:
```

Comment:
Having to handle `List` types here is not ideal.

Reply:
Do you mean define our own `List` type?

Reply:
Yeah, maybe it might worth clearly defining such custom types separately for the model. And since they're specific to the model they won't be under common ones:
 `from outlines.types import Choice, Json`

 But instead from somewhere like:
 `from outlines.gemini.types import List`?

Considering the variations of the models, it seems it might be quite common situation to manage the difference between them all. What do you think?

Reply:
Yeah, extending the common when we are sure it's a good addition is ok, I wonder how to handle purely custom ones, which we wouldn't want to support for any other model but this particular one. For that we'd need some good examples, if List is out to be a purely custom one.